docs/user/exercises/programming-exercise-AI-proof-guidelines.inc

```diff
@@ -0,0 +1,491 @@
+          }
+          // Sort the shopping list by value in descending order
+          Arrays.sort(shoppingList, (a, b) -> b.getValue() - a.getValue());
+          ...
+      }
+
+   Example solution:
+
+   .. code:: java
+
+      public Item[] fillBagMax() {
+          if (checkShoppingListExists()) {
+              //create a new array without null item to aplly the
+              //sorting algorithm
+              int size = 0;
+              for (int i = 0; i < shoppingList.length; i++) {
+                  if (checkItemExists(i)) {
+                      size++;
+                  }
+              }
+              Item[] listWithoutNull = new Item[size];
+              int nextInShoppingList = 0;
+
+              for (int i = 0; i < listWithoutNull.length; i++) {
+                  // Move the pointer one further, if a spot on the list
+                  //is not filled.
+                  while (shoppingList[nextInShoppingList] == null) {
+                      nextInShoppingList++;
+                  }
+                  listWithoutNull[i] = shoppingList[nextInShoppingList++];
+              }
+              ...
+      }
+
+   .. note::
+      The java standard methods which can be used to solve specific
+      tasks need to be excluded in the test cases and feedback has to be provided to the student.
+      e.g. The method Arrays.sort is not allowed to be used.
+
+
+-  **Continuous feedback:** Exercise instructions should not be there to just give a
+   thorough step by step guide of the task. Have students deduce, using specific test cases,
+   what might be missing in the code.
+   This promotes critical thinking and resembles a more real world scenario:
+
+   -  null-checks
+
+   -  check existence of list
+
+   Example shows that AI disregards to check if shopping list exists:
+
+   Generated solution:
+
+   .. code:: java
+
+      public int search(String itemName) {
+         for (int i = 0; i < shoppingList
+                     .length; i++) {
+            if (shoppingList[i].getName()
+                     .equals(itemName)) {
+            return i;
+            }
+         }
+         return -1;
+      }
+
+   Example solution:
+
+   .. code:: java
+
+      public int search(String itemName) {
+      //check whether the shoppingList exists
+         if (!checkShoppingListExists()) {
+            return -1;
+         }
+
+         //the rest is the same as
+         //the AI solution
+
+      }
+
+-  **Split Up Instructions for Improved Engagement:** To prevent
+   students from blindly copying and pasting, it's essential to
+   distribute the instructions strategically. Ideally, around 80%
+   of the guidelines should be placed within the problem
+   statement, leaving a minimum of 20% to be incorporated
+   directly into the template code, marked clearly with "TODO"
+   annotations. This approach ensures students interact more
+   deeply with the code while also referencing the overarching
+   problem statement.
```

Comment:
This is true, but is also easily circumvented by providing the template code to ChatGPT as well. This only really works well for complex exercises.

Reply:
True. It just prevents a blind copy and paste attempt from the problem statement. A more thorough look into the exercises is required. Thats helps to understand the exercise and maybeencourages the students to solve on their own instead of only relying on generated code. A lot can be circumvented, but it is more about encouraging a self-implementation and understanding the problem instead of blind copy and paste attempts.